src/passes/MergeFunctions.cpp

```diff
@@ -0,0 +1,609 @@
+  }
+}
+
+// Find the set of parameters which are required to merge the functions in the
+// class Returns false if unable to derive parameters.
+bool EquivalentClass::deriveParams(Module* module,
+                                   std::vector<ParamInfo>& params,
+                                   bool isCallIndirectionEnabled) {
+  struct DeepValueIterator {
+    SmallVector<Expression**, 10> tasks;
+
+    DeepValueIterator(Expression** root) : tasks() { tasks.push_back(root); }
+
+    void operator++() {
+      ChildIterator it(*tasks.back());
+      tasks.pop_back();
+      for (Expression*& child : it) {
+        tasks.push_back(&child);
+      }
+    }
+
+    Expression*& operator*() { return *tasks.back(); }
+    bool empty() { return tasks.empty(); }
+  };
+
+  if (primaryFunction->imported()) {
+    return false;
+  }
+  DeepValueIterator primaryIt(&primaryFunction->body);
+  std::vector<DeepValueIterator> siblingIterators;
+  // Skip the first function, as it is the primary function.
+  for (auto func = functions.begin() + 1; func != functions.end(); ++func) {
+    siblingIterators.emplace_back(&(*func)->body);
+  }
+
+  for (; !primaryIt.empty(); ++primaryIt) {
+    Expression*& primary = *primaryIt;
+    ConstDiff diff;
+    Literals values;
+    std::vector<Name> names;
+
+    bool isAllSame = true;
+    if (auto* primaryConst = primary->dynCast<Const>()) {
+      values.push_back(primaryConst->value);
+      for (auto& it : siblingIterators) {
+        Expression*& sibling = *it;
+        ++it;
+        if (auto* siblingConst = sibling->dynCast<Const>()) {
+          isAllSame &= primaryConst->value == siblingConst->value;
+          values.push_back(siblingConst->value);
+        } else {
+          WASM_UNREACHABLE(
+            "all sibling functions should have the same instruction type");
+        }
+      }
+      diff = values;
+    } else if (isCallIndirectionEnabled && primary->is<Call>()) {
+      auto* primaryCall = primary->dynCast<Call>();
+      names.push_back(primaryCall->target);
+      for (auto& it : siblingIterators) {
+        Expression*& sibling = *it;
+        ++it;
+        if (auto* siblingCall = sibling->dynCast<Call>()) {
+          isAllSame &= primaryCall->target == siblingCall->target;
+          names.push_back(siblingCall->target);
+        } else {
+          WASM_UNREACHABLE(
+            "all sibling functions should have the same instruction type");
+        }
+      }
+      diff = names;
+    } else {
+      // Skip non-constant expressions, which are ensured to be the exactly
+      // same.
+      for (auto& it : siblingIterators) {
+        // Sibling functions in a class should have the same instruction type.
+        assert((*it)->_id == primary->_id);
+        ++it;
+      }
+      continue;
+    }
+    // If all values are the same, skip to parameterize it.
+    if (isAllSame) {
+      continue;
+    }
+    // If the derived param is already in the params, reuse it.
```

Comment:
I don't follow this. Perhaps a small example?

Reply:
Added an example for this case